packages/eas-cli/src/credentials/ios/appstore/ensureAppExists.ts

```diff
@@ -215,3 +216,120 @@ export async function ensureAppExistsAsync(
+          `App name "${props.name}" is already taken. Using generated name "${generatedName}" which can be changed later from https://appstoreconnect.apple.com.`
+        );
+        // Sanitize the name and try again.
+        return await createAppAsync(
+          context,
+          {
+            ...props,
+            name: generatedName,
+          },
+          retryCount + 1
+        );
+      };
+
+      if (isAppleError(error)) {
+        // New error class that is thrown when the name is already taken but belongs to you.
+        if (
+          error.data.errors.some(
+            e =>
+              e.code === 'ENTITY_ERROR.ATTRIBUTE.INVALID.DUPLICATE.SAME_ACCOUNT' ||
+              e.code === 'ENTITY_ERROR.ATTRIBUTE.INVALID.DUPLICATE.DIFFERENT_ACCOUNT'
+          )
+        ) {
+          return await handleDuplicateNameErrorAsync();
+        }
+      }
+
+      if ('code' in error && typeof error.code === 'string') {
+        if (
+          // Name is invalid
+          error.code === 'APP_CREATE_NAME_INVALID'
+          // UnexpectedAppleResponse: An attribute value has invalid characters. - App Name contains certain Unicode symbols, emoticons, diacritics, special characters, or private use characters that are not permitted.
+          // Name is taken
+        ) {
+          const sanitizedName = sanitizeName(props.name);
+          if (sanitizedName === props.name) {
+            throw error;
+          }
+          Log.warn(
+            `App name "${props.name}" contains invalid characters. Using sanitized name "${sanitizedName}" which can be changed later from https://appstoreconnect.apple.com.`
+          );
+          // Sanitize the name and try again.
+          return await createAppAsync(
+            context,
+            {
+              ...props,
+              name: sanitizedName,
+            },
+            retryCount + 1
+          );
+        }
+
+        if (
+          // UnexpectedAppleResponse: The provided entity includes an attribute with a value that has already been used on a different account. - The App Name you entered is already being used. If you have trademark rights to
+          // this name and would like it released for your use, submit a claim.
+          error.code === 'APP_CREATE_NAME_UNAVAILABLE'
+        ) {
+          return await handleDuplicateNameErrorAsync();
+        }
+      }
+    }
+
+    throw error;
+  }
+}
+
+function isAppleError(error: any): error is {
+  data: {
+    errors: {
+      id: string;
+      status: string;
+      /** 'ENTITY_ERROR.ATTRIBUTE.INVALID.INVALID_CHARACTERS' */
+      code: string;
+      /** 'An attribute value has invalid characters.' */
+      title: string;
+      /** 'App Name contains certain Unicode symbols, emoticons, diacritics, special characters, or private use characters that are not permitted.' */
+      detail: string;
+    }[];
+  };
+} {
+  return 'data' in error && 'errors' in error.data && Array.isArray(error.data.errors);
```

Comment:
We can try to validate this data against a scheme by using `zod` or `joi` instead of writing if statements manually